homeassistant/components/thethingsnetwork/__init__.py

```diff
@@ -33,10 +34,57 @@
 
 async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
     """Initialize of The Things Network component."""
-    conf = config[DOMAIN]
-    app_id = conf.get(CONF_APP_ID)
-    access_key = conf.get(CONF_ACCESS_KEY)
 
-    hass.data[DATA_TTN] = {TTN_ACCESS_KEY: access_key, TTN_APP_ID: app_id}
+    if DOMAIN in config:
+        ir.async_create_issue(
+            hass,
+            DOMAIN,
+            "manual_migration",
+            breaks_in_ha_version="2021.10.0",
+            is_fixable=False,
+            severity=ir.IssueSeverity.ERROR,
+            translation_key="manual_migration",
+            translation_placeholders={"domain": DOMAIN},
+        )
+
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Establish connection with The Things Network."""
+
+    _LOGGER.debug(
+        "Set up %s at %s",
+        entry.data[CONF_APP_ID],
+        entry.data.get(CONF_HOSTNAME, TTN_API_HOSTNAME),
+    )
+
+    # Create coordinator to fetch TTN updates
+    coordinator = TTNCoordinator(hass, entry)
+    hass.data.setdefault(DOMAIN, {}).setdefault(entry.entry_id, {})[
+        ENTRY_DATA_COORDINATOR
+    ] = coordinator
```

Comment:
You're not storing anything else than the coordinator in the entry data, so you can just only store that
```suggestion
    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = coordinator
```

Please have the `coordinator.async_config_entry_first_refresh` before setting it to `hass.data` this way we only store objects that work.